<h5 align="center">Metody Obliczeniowe w Nauce i Technice </h5>
<h6 align="center">Labolatorium 5</h6>
<h6 align="center">Równania nieliniowe</h6>
<h6 align="center">7 kwietnia 2020</h6>

In [21]:
from mpmath import cos,cosh,sec,sin,sinh,exp,tan,fabs,mpf,log,mp,pi

<b>Zdefiniowanie funkcji tesowych</b> <br>

\begin{equation*}
f_{1}(x) = cos(x)cosh(x) - 1, [\frac{3}{2} \pi,2 \pi]
\end{equation*}
\begin{equation*}
f_{2}(x) = \frac{1}{x} - tan(x), [0, \frac{ \pi }{2}]
\end{equation*}
\begin{equation*}
f_{3}(x) = 2^{~-x} + e^{x} + 2cos(x) - 6, [1,3]
\end{equation*}

In [22]:
f1 = lambda x : cos(x)*cosh(x) - 1
f2 = lambda x : 1/x  - tan(x)
f3 = lambda x : 2**(-x) + exp(x) + 2*cos(x) - 6

left_f1 = mpf(3/2*pi)
right_f1 = mpf(2*pi)

left_f2 = mpf(0)
right_f2 = mpf(pi/2)

left_f3 = mpf(1)
right_f3 = mpf(3)

<b>Zadanie 1 Metoda Bisekcji</b> <br><hr>
Pieszemy funkcję realizującą metodę bisekcji dla danej funkcji f w oparciu o arytmetykę o zmiennej precyzji (mpmath.mpf). Funkcja przyjmuje następujące argumenty:
* Minimalną precyzję obliczeń (liczba cyfr znaczących)
* Krańce przedziału
* Błąd bezwzględny obliczeń

Funkcja ma zwracać wyznaczone miejsce zerowe, wyliczoną wartość w miejscu zerowym oraz liczbę iteracji potrzebną do uzyskania określonej dokładności. Testujemy działanie metody dla funkcji podanych na początku instrukcji. Jaka liczba iteracji jest potrzebna do uzyskania bezwzględnej dokładności rzędu: 10−7, 10−15, 10−33? W jaki sposób możemy obliczyć k pierwszych dodatnich pierwiastków funkcji f1(x)?

* Funkcja realizująca metode bisekcji

In [23]:
def bisection(min_precision,left,right,absolute_error,f):

    # ustawienie precyzji
    mp.dps = min_precision
    # zfracany wnik
    x = mpf(0.0)
    # liczba krokow
    k = 0

    left = mpf(left)
    right = mpf(right)
    absolute_error = mpf(absolute_error)

    while fabs(left - right) > absolute_error:

        x = mpf((left + right)/2)
        if fabs(f(x)) <= absolute_error:
            break
        elif mpf(f(left))*mpf(f(x)) < mpf(0):
            right = x
        elif mpf(f(right))*mpf(f(x)) < mpf(0):
            left = x
        k += 1

    print("Miejsce zerowe",x)
    print("Wartosc funkcji",f(x))
    print("Liczba krokow",k)
    print()

* Testy dla błędu bezwzględnego 10^-7:

In [24]:
# dla f1
bisection(7,left_f1,right_f1,mpf(10**(-7)),f1)

#  dla f2
bisection(7,left_f2 +mpf(10**(-7)),right_f2,mpf(10**(-7)),f2)

#  dla f3
bisection(7,left_f3,right_f3,mpf(10**(-7)),f3)

Miejsce zerowe 4.730041
Wartosc funkcji -9.164214e-7
Liczba krokow 25

Miejsce zerowe 0.8603336
Wartosc funkcji -7.450581e-8
Liczba krokow 23

Miejsce zerowe 1.829384
Wartosc funkcji 0.0
Liczba krokow 22



* Testy dla błędu bezwzględnego 10^-15:

In [25]:
#  dla f1
bisection(15,left_f1,right_f1,mpf(10**(-15)),f1)

#  dla f2
bisection(15,left_f2+mpf(10**(-15)),right_f2,mpf(10**(-15)),f2)

#  dla f3
bisection(15,left_f3,right_f3,mpf(10**(-15)),f3)

Miejsce zerowe 4.7300407448627
Wartosc funkcji 3.574918139293e-14
Liczba krokow 51

Miejsce zerowe 0.86033358901938
Wartosc funkcji 0.0
Liczba krokow 50

Miejsce zerowe 1.82938360193385
Wartosc funkcji -8.88178419700125e-16
Liczba krokow 47



* Testy dla błędu bezwzględnego 10^-33:

In [26]:
#  dla f1
bisection(33,left_f1,right_f1,mpf(10**(-33)),f1)

#  dla f2
bisection(33,left_f2+mpf(10**(-7)),right_f2,mpf(10**(-33)),f2)

#  dla f3
bisection(33,left_f3,right_f3,mpf(10**(-33)),f3)

Miejsce zerowe 4.73004074486270402602404810083388
Wartosc funkcji -4.33334237487128066937595037081609e-33
Liczba krokow 111

Miejsce zerowe 0.860333589019379762483893424137662
Wartosc funkcji 9.62964972193617926527988971292464e-34
Liczba krokow 110

Miejsce zerowe 1.82938360193384881713621294681415
Wartosc funkcji 7.70371977754894341222391177033971e-34
Liczba krokow 110



* <b>Wnioski:</b><br> Do uzyskania potrzebnej dokładności wyagane jest odpowiednio około 23,50, oraz 110 kroków. Do lewego krańca przedzialu funkcji nr 2 dodajemy wartość błędu bezwzględnego. W celu uniknięcia dzielenia przez 0 (0 nie należy do dzeidziny tej funkcji).Liczba kroków rośnie logarytmicznie według wzoru:


\begin{equation*}
n = \frac{log\frac{b-a}{\epsilon }}{log 2}
\end{equation*}

Można to prosto uzasadnić, w przedziale [a,b] mieści się (b-a)/ε przedziałów długości ε. Przekształcając powyższy wzór otrzymujemy:

\begin{equation*}
2^{n} =  \frac{b-a}{\epsilon}
\end{equation*}

Jako, że za każdym razem dzieliny nowy przedział na dwa to musimy wykonać n kroków, aby uzyskać przedział długości ε. Patrząc z drugiej strony, mamy przedziały długości ε, za każdym razem podwajamy ich ilosć, żeby ostatecznie uzyskać ich ilość odpowiadającą przedziałowi [a,b], więc (b-a)/ε małych przedziałów. Możemy tylko podwajać ilosć przedziałów, wiec ich liczba w każdym kroku to 2^n.

* <b>W jaki sposób możemy obliczyć k pierwszych dodatnich pierwiastków funkcji f1(x)?</b>

Funckja nr 1 jest funkcją okresową, znamy więc przedziały w których występują miejsca zerowe. Wystarczy więc k razy wywołać bisekcje dla kolejncyh przedziałów odległych o pi.

<b>Zadanie 2 Metoda Newtona</b> <br><hr>
Pieszemy funkcję  realizującą metodę Newtona w oparciu o arytmetykę o zmiennej precyzji (mpmath.mpf). Funkcja ma wykorzystywać dwa kryteria stopu: 
* maksymalną liczbę iteracji
* moduł różnicy kolejnych przybliżeń mniejszy od danej wartości ε

Oprócz przybliżonej wartości pierwiastka funkcja ma zwrócić liczbę iteracji potrzebną do uzyskania określonej dokładności ε. Testujemy działanie metody dla funkcji podanych na początku sprawozdania (dodatkowo dostępne pochodne tych funkcji). Porównujemy zbieżność metody ze zbieżnością uzyskaną dla metody bisekcji.


* Zdefiniowanie pochodnych funkcji wejsciowych

In [27]:
df1 = lambda x : -cosh(x)*sin(x) + cos(x)*sinh(x)
df2 = lambda x : -1/(x**2) - sec(x)**2
df3 = lambda x : exp(x) - 2**(-x)*log(2) - 2*sin(x)

* Funkcja realuzująca metode Newtona

In [45]:
def newton_method(min_precision,left,right,epsilon,f,df,max_iterations):
    mp.dps = min_precision
    k = 0
    start_point = mpf(left)
    x = mpf(right)
    epsilon = mpf(epsilon)
    while k < max_iterations:

        x = start_point - mpf(f(start_point))/mpf(df(start_point))

        if  fabs(start_point - x) <= epsilon:
            break
        start_point = x
        k += 1

    print("Miejsce zerowe",x)
    print("Wartosc funkcji",f(x))
    print("Liczba krokow",k)
    print()

* Testy dla błędu bezwzględnego 10^-7:

In [46]:
#  dla f1
newton_method(7,left_f1,right_f1,mpf(10**(-7)),f1,df1,200)

#  dla f2
newton_method(7,left_f2+mpf(10**(-6)),right_f2,mpf(10**(-7)),f2,df2,200)

#  dla f3
newton_method(7,left_f3,right_f3,mpf(10**(-7)),f3,df3,200)

Miejsce zerowe 4.730041
Wartosc funkcji -9.164214e-7
Liczba krokow 3

Miejsce zerowe 0.8603336
Wartosc funkcji 0.0
Liczba krokow 22

Miejsce zerowe 1.829384
Wartosc funkcji 0.0
Liczba krokow 7



* Testy dla błędu bezwzględnego 10^-15:

In [47]:
#  dla f1
newton_method(15,left_f1,right_f1,mpf(10**(-15)),f1,df1,200)

#  dla f2
newton_method(15,left_f2+mpf(10**(-14)),right_f2,mpf(10**(-15)),f2,df2,200)

#  dla f3
newton_method(15,left_f3,right_f3,mpf(10**(-15)),f3,df3,200)

Miejsce zerowe 4.7300407448627
Wartosc funkcji -1.53210777398272e-14
Liczba krokow 4

Miejsce zerowe 0.86033358901938
Wartosc funkcji 0.0
Liczba krokow 50

Miejsce zerowe 1.82938360193385
Wartosc funkcji 0.0
Liczba krokow 8



* Testy dla błędu bezwzględnego 10^-33:

In [48]:
#  dla f1
newton_method(33,left_f1,right_f1,mpf(10**(-33)),f1,df1,200)

#  dla f2
newton_method(33,left_f2+mpf(10**(-32)),right_f2,mpf(10**(-33)),f2,df2,200)

#  dla f3
newton_method(33,left_f3,right_f3,mpf(10**(-33)),f3,df3,200)

Miejsce zerowe 4.73004074486270402602404810083388
Wartosc funkcji -4.33334237487128066937595037081609e-33
Liczba krokow 5

Miejsce zerowe 0.860333589019379762483893424137662
Wartosc funkcji -1.92592994438723585305597794258493e-34
Liczba krokow 111

Miejsce zerowe 1.82938360193384881713621294681415
Wartosc funkcji 0.0
Liczba krokow 9



* <b>Wnioski</b><br>Do lewego krańca przedzialu funkcji nr 2 dodajemy wartość błędu bezwzględnego, aby uniknąć dzielenia przez 0 (0 nie należy do dzeidziny tej funkcji). Do uzyskania potrzebnej dokładności wymagane jest odpowiednio 7,8,9 kroków jet to o wiele mniej niż w przypadku metody bisekcji. Wadą tego rozwiązania jest konieczność znania pochodnej funckji.

<b>Zadanie 3 Metoda siecznych</b> <br><hr>
Pieszemy funkcję  realizującą metodę siecznych w oparciu o arytmetykę o zmiennej precyzji (mpmath.mpf). Funkcja ma wykorzystywać dwa kryteria stopu: 
* maksymalną liczbę iteracji
* moduł różnicy kolejnych przybliżeń mniejszy od danej wartości ε

Testujemy działanie metody dla funkcji podanych na początku sprawozdania. Porównujemy zbieżność metody ze zbieżnością uzyskaną dla metody bisekcji oraz metody Newtona.

* Funckja realizujaca metodę siecznych

In [49]:
def euler_method(min_precision,left,right,epsilon,f,max_iterations,k):
    mp.dps = min_precision
    left = mpf(left)
    right = mpf(right)
    epsilon = mpf(epsilon)
    k = 0
    while max_iterations > 0 and fabs(left - right) > epsilon:
        max_iterations -= 1
        k += 1
        x0 = left - f(left)*(left - right)/(f(left) - f(right))
        right = left
        left = x0
        if f(x0) < epsilon:
            break
    print("Miejsce zerowe",x0)
    print("Wartosc funkcji",f(x0))
    print("Liczba krokow",k)
    print()

* Testy dla błędu bezwzględnego 10^-7:

In [55]:
newton_method(7,left_f1,right_f1,mpf(10**(-7)),f1,df1,200)

#  dla f2
newton_method(7,left_f2 + mpf(10**(-6)),right_f2,mpf(10**(-7)),f2,df2,200)

#  dla f3
newton_method(7,left_f3,right_f3,mpf(10**(-7)),f3,df3,200)

Miejsce zerowe 4.730041
Wartosc funkcji -9.164214e-7
Liczba krokow 3

Miejsce zerowe 0.8603336
Wartosc funkcji 0.0
Liczba krokow 22

Miejsce zerowe 1.829384
Wartosc funkcji 0.0
Liczba krokow 7



* Testy dla błędu bezwzględnego 10^-15:

In [54]:
#  dla f1
newton_method(15,left_f1,right_f1,mpf(10**(-15)),f1,df1,200)

#  dla f2
newton_method(15,left_f2 + mpf(10**(-14)),right_f2,mpf(10**(-15)),f2,df2,200)

#  dla f3
newton_method(15,left_f3,right_f3,mpf(10**(-15)),f3,df3,200)

Miejsce zerowe 4.7300407448627
Wartosc funkcji -1.53210777398272e-14
Liczba krokow 4

Miejsce zerowe 0.86033358901938
Wartosc funkcji 0.0
Liczba krokow 50

Miejsce zerowe 1.82938360193385
Wartosc funkcji 0.0
Liczba krokow 8



* Testy dla błędu bezwzględnego 10^-33:

In [53]:
#  dla f1
newton_method(33,left_f1,right_f1,mpf(10**(-33)),f1,df1,200)

#  dla f2
newton_method(33,left_f2 + mpf(10**(-32)),right_f2,mpf(10**(-33)),f2,df2,200)

#  dla f3
newton_method(33,left_f3,right_f3,mpf(10**(-33)),f3,df3,200)

Miejsce zerowe 4.73004074486270402602404810083388
Wartosc funkcji -4.33334237487128066937595037081609e-33
Liczba krokow 5

Miejsce zerowe 0.860333589019379762483893424137662
Wartosc funkcji -1.92592994438723585305597794258493e-34
Liczba krokow 111

Miejsce zerowe 1.82938360193384881713621294681415
Wartosc funkcji 0.0
Liczba krokow 9



* <b>Wnioski</b><br>Do lewego krańca przedzialu funkcji nr 2 dodajemy wartość błędu bezwzględnego, aby uniknąć dzielenia przez 0 (0 nie należy do dzeidziny tej funkcji). Do uzyskania potrzebnej dokładności wymagane jest odpowiednio bardzo mało kroków. W przypadku funkcji nr 2 liczba kroków jest bardzo duża, gorsza niż w przypadku metod Newtona i bisekcji. Metoda siecznych jest więc dobra, ale nie nadaje się do badania każdej funkcji. W przypadku funkcji f1 i f3 metoda ta potrzebowała mniej kroków do osiągnięcia tej samej precyzji. Jednak dla funkcji f2 ta metoda zawodzi.

<h3>Podsumowanie</h3><hr>

<h6>Metoda bisekcji</h6><br>
 Metoda bisekcji jest zbieżna liniowo z ilorazem 1/2. Choć ta zbieżność nie jest imponująca, bisekcja ma kilka istotnych zalet. Oprócz jej prostoty, należy podkreślić fakt, że bisekcja jest w pewnym sensie uniwersalna. Jeśli tylko dysponujemy dwoma punktami a i b takimi, że f przyjmuje w nich wartości przeciwnych znaków, to metoda bisekcji z pewnością znajdzie miejsce zerowe metoda jest zawsze zbieżna

<b>Zalety:</b>
- metoda jest zawsze zbieżna
-  przedział, w którym znajduje się pierwiastek jest zawsze połowiony

<b>Wady</b>
- metoda jest wolnozbieżna
- jeżeli pierwiastek odgadnięty jest bliski rzeczywistemu to szybkość maleje
- Jeżeli funkcja f(x) jest taka, że dotyka osi OX (jak w f(x) = x^2) to nie można znaleźć pierwiastka metodą bisekcji

<h6>Metoda Newtona</h6><br>
Zaletą metody Newtona jest jej szybka zbieżność (w algorytmie liczba cyfr znaczących w kolejnych przybliżeniach podwaja się). Wadą jej jest fakt, iż zbieżność nie musi zawsze zachodzić.Drugą wadą jest to, że trzeba znać pochodną funkcji w punkcie. W wielu przypadkach metoda bywa rozbieżna, kiedy punkt startowy jest zbyt daleko od szukanego pierwiastka równania.


<h6>Metoda Siecznych</h6><br>
Metoda siecznych ma tę zaletę, że do wykonania interpolacji za jej pomocą niepotrzebna jest znajomość pochodnej danej funkcji. Wadą jest to, że metoda ta nie zawsze jest zbieżna.

<b>Źródła</b><br>
http://wazniak.mimuw.edu.pl/index.php?title=MN02
https://pl.wikipedia.org/wiki/Metoda_r%C3%B3wnego_podzia%C5%82u
https://pl.wikipedia.org/wiki/Metoda_siecznych
https://pl.wikipedia.org/wiki/Metoda_r%C3%B3wnego_podzia%C5%82u
http://home.agh.edu.pl/~zak/downloads/MN4-2012.pdf